### ISSUE TITLE
Inconsistency of loadPixels method in different browsers

### ISSUE THREAD

**Issue by N4M3L355** [processing/p5.js#5064]
#### Most appropriate sub-area of p5.js?

- [ ] Accessibility (Web Accessibility)
- [ ] Build tools and processes
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Friendly error system
- [x] Image
- [ ] IO (Input/Output)
- [ ] Localization
- [ ] Math
- [ ] Unit Testing
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Other (specify if possible)

#### Details about the bug:

- p5.js version: 1.2.0 
- Web browser and version: does not give incorrect numbers in Firefox 87.0b3, gives correct numbers in Chrome 87.0.4280.88 
- Operating System: Linux Pop!_OS 20.04 
- Steps to reproduce this:

This code assigns certain values to pixels, then updatesPixels, then loadsPixels. The pixels aren't the same as the value assigned to them previously. The problem is present in Firefox and is not present in Chrome. Is there any workaround for this values to persist correctly in all browsers?
```javascript
function setup() {
  createCanvas(2**6,2**6);
  noSmooth();             //this is just to be sure smoothing is not the reason this happens
  pixelDensity(1);        //to force 1 to 1 display density
  loadPixels();
  testArray = [255, 255, 255, 254, 255, 255, 255, 253]
  testArray.forEach((x,i) => pixels[i] = x);
  updatePixels();
  loadPixels();
  console.log(pixels.slice(0,8));  //expected: [255, 255, 255, 254, 255, 255, 255, 253] 
                                   //got:      [254, 254, 254, 254, 255, 255, 255, 253]
}                                  //             ^    ^    ^

```
I've also used p5js editor for live preview — https://editor.p5js.org/adamkral7560/sketches/5kAkDB4XS.

**Comment by limzykenneth** [comment]
`loadPixels()` doesn't need to be called after `updatePixels()` here. Removing the second `loadPixels()` in your example gives the expected output.

**Comment by N4M3L355** [comment]
I may have defined my problem unclearly; let me rephrase it. When I assign pixels certain values, save them to the picture, and then try to load values previously assigned to them, they are not the same. Maybe I can further clarify with the scenario in which I've found this bug: I have a project where I wanted to harness steganography to include some text in the resulting image data. Let's say a function saves information to image data (by changing pixels[] and calling updatePixels). I now want to retrieve them from the image (by using loadPixels and accessing pixels[] values), but I am unsuccessful, as I am unable to update pixels so that the RGBA values will stay the same as assigned.

**Comment by limzykenneth** [comment]
You don't have to call `loadPixels()` more than once. If you have called it before, edit the `pixels` array then `updatePixels()`, in the other parts of your code you can still access `pixels` without calling `loadPixels()`. `loadPixels()` should only ever be called once to populate the `pixels` array as it is initialized empty, it is not used to update the `pixels` object since your code is supposed to be the one updating it.

**Comment by N4M3L355** [comment]
Yes, I understand, but I wanted to show that when you call loadPixels in Firefox, the actual color of that pixel on canvas won't be set correctly. In my scenario, if I do the following steps:
1. save to pixels[] value (with setting pixels[] and then calling updatePixels())
2. export the png from canvas (with save())
3. load the png to canvas (with loadImage())
4. loadPixels
Then the result will vary with the browser. The first and second steps are not even needed, you can try to loadPixels from this image: 
![index](https://user-images.githubusercontent.com/16424778/109438738-7c264f80-7a2b-11eb-8f89-191fbdc9f469.png)
Or you can use this p5js editor demo: https://editor.p5js.org/adamkral7560/sketches/RhMb_7wGW
First pixel is 255,255,255,254 — the p5js pixels[] report it is 254,254,254,254 (in firefox, chrome is correct).
I'm sorry if the first description was too complicated, I thought that it will be the best example of a minimal case.

**Comment by vsquareg** [comment]
This seems to be caused by the PreMultiplied Alpha used by browsers, probably a rounding error when multiplying the alpha value. It has been a known issue specifically with Firefox on macOS.

Some links you can follow to read more about this ([Firefox canvas getImageData...](https://stackoverflow.com/questions/51089924/firefox-canvas-getimagedata-values-differ-from-original-image-pixel-values)) ([Why does HTML Canvas getImageData() not return the exact...](https://stackoverflow.com/questions/4309364/why-does-html-canvas-getimagedata-not-return-the-exact-same-values-that-were-j)).

It has also been mentioned specifically as a note in the HTML5 specs as well ([link](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-putimagedata)).

### Solution?

There's nothing that can be done within the `loadPixels` or `updatePixels` function to mitigate this, it is caused inherently by the rendering engine of the browser.

However, for the purposes of steganography, as it is a standard suggestion, you can use alpha values as 255 (255/255 = 1), so that the rounding error would be avoided. This is the easiest way to guarantee consistency.

**Comment by lmccart** [comment]
It looks like the references @vsquareg linked are correct. It seems we can't fix this within p5 so I will close this issue. Thanks for reporting!
